File: day15part2.clj
(println "Java" (System/getProperty "java.version"))
;(def starting [0 3 6])
(def starting [0,5,4,1,10,14,7])

(def target 30000000)

(def starting-spoken-before
  (reduce
    (fn [acc [n index]]
      (assoc acc index (list n n)))
    (vec (repeat target nil))
    (map vector (drop 1 (range)) starting)))

(defn spoken-before?
  [m k]
  (>= (count (get m k)) 2))

(defn update-indices
  [m k n]
  (let [indices (get m k)
        new-indices (if (nil? indices)
                      (list n n)
                      (list n (first indices)))]
    (assoc m k new-indices)))

(loop [spoken-before starting-spoken-before
       last-spoken (last starting)
       index (inc (count starting))]
  (if (= target (dec index))
    (println "Answer: " last-spoken)
    (let [next-n (reduce - (get spoken-before last-spoken))]
      (recur (update-indices spoken-before next-n index) next-n (inc index)))))

File: Day15Part2.java
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

class Day15Part2 {

    public static void main(String[] args) {

        int target = 30_000_000;
        //int target = 2020;
        Map<Integer,int[]> spoken = new HashMap<>();
        List<Integer> starting = Arrays.asList(0,5,4,1,10,14,7);

        int i = 0;
        for (int n : starting) {
            spoken.put(n, new int[] {++i, i});
        }

        int index = starting.size() + 1;
        int lastSpoken = starting.get(starting.size() - 1);
        while (index <= target) {
            //System.out.println(lastSpoken);
            int[] indices = spoken.get(lastSpoken);
            int nextNum = indices[0] - indices[1];
            int[] nextNumIndices = spoken.get(nextNum);
            int prevIndex = nextNumIndices == null ? index : nextNumIndices[0];
            spoken.put(nextNum, new int[] {index, prevIndex});
            lastSpoken = nextNum;
            index++;
        }
        System.out.println(lastSpoken);
    }
}
